dbos/_core/system_database.py

```diff
@@ -0,0 +1,1734 @@
+        )
+
+        # If the system database does not already exist, create it
+        postgres_db_url = sa.URL.create(
+            "postgresql+psycopg",
+            username=config["database"]["username"],
+            password=config["database"]["password"],
+            host=config["database"]["hostname"],
+            port=config["database"]["port"],
+            database="postgres",
+        )
+        engine = sa.create_engine(postgres_db_url)
+        with engine.connect() as conn:
+            conn.execution_options(isolation_level="AUTOCOMMIT")
+            if not conn.execute(
+                sa.text("SELECT 1 FROM pg_database WHERE datname=:db_name"),
+                parameters={"db_name": sysdb_name},
+            ).scalar():
+                conn.execute(sa.text(f"CREATE DATABASE {sysdb_name}"))
+        engine.dispose()
+
+        system_db_url = sa.URL.create(
+            "postgresql+psycopg",
+            username=config["database"]["username"],
+            password=config["database"]["password"],
+            host=config["database"]["hostname"],
+            port=config["database"]["port"],
+            database=sysdb_name,
+        )
+
+        # Run a schema migration for the system database
+        migration_dir = os.path.join(
+            os.path.dirname(os.path.realpath(__file__)), "migrations"
+        )
+        alembic_cfg = Config()
+        alembic_cfg.set_main_option("script_location", migration_dir)
+        alembic_cfg.set_main_option(
+            "sqlalchemy.url", system_db_url.render_as_string(hide_password=False)
+        )
+        command.upgrade(alembic_cfg, "head")
+
+        self.notification_conn: Optional[psycopg.Connection] = None
+        self.notifications_map: Dict[str, threading.Condition] = {}
+        self.workflow_events_map: Dict[str, threading.Condition] = {}
+
+        # Initialize the workflow status and inputs buffers
+        self._workflow_status_buffer: Dict[str, WorkflowStatusInternal] = {}
+        self._workflow_inputs_buffer: Dict[str, str] = {}
+        # Two sets for tracking which single-transaction workflows have been exported to the status table
+        self._exported_temp_txn_wf_status: Set[str] = set()
+        self._temp_txn_wf_ids: Set[str] = set()
+        self._is_flushing_status_buffer = False
+
+        # Now we can run background processes
+        self._run_background_processes = True
+
+        # Create a connection pool for the system database
+        self.sync_engine = sa.create_engine(
+            system_db_url, pool_size=20, max_overflow=5, pool_timeout=30
+        )
+        self.async_engine = create_async_engine(
+            system_db_url, pool_size=20, max_overflow=5, pool_timeout=30
+        )
+
+    # Destroy the pool when finished
+    def destroy(self) -> None:
+        asyncio.run(self.destroy_async())
+
+    async def destroy_async(self) -> None:
+        self.wait_for_buffer_flush()
```

Comment:
I suppose this is not ideal in an async method but it is destroy and not a big deal?

Reply:
That is, if you were in an actual async context and not under asyncio.run in a thread, the wait would never finish because the coroutine never yields.

Reply:
No, it's not ideal to have asyncio.run anywhere in our code. However, the `AsyncEngine` only has an async `dispose` method. So our options for a sync `destroy` method are:

* `asyncio.run(self.async_engine.dispose())`
* `self.async_engine.sync_engine.dispose()`

SQLAlchemy's async objects are basically wrappers around their sync objects w/ async methods implemented using [`greenlet`](https://greenlet.readthedocs.io/en/latest/), a lightweight coroutine library. The implementaiton of [`AsyncEngine.dispose`](https://github.com/sqlalchemy/sqlalchemy/blob/fe03298593e59e5d40de81e7f864690e73930666/lib/sqlalchemy/ext/asyncio/engine.py#L1126-L1149) is literally just `await greenlet_spawn(self.sync_engine.dispose, close=close)`. `sync_engine` is exposed as a public field of `AsyncEngine` in order to be used as a target in the [SQL Alchemy event system](https://docs.sqlalchemy.org/en/20/orm/extensions/asyncio.html#using-events-with-the-asyncio-extension). You're not *supposed* to call methods on it directly AFAIK

Reply:
I meant wait_for_buffer_flush specifically.  That will be a blocking wait, which will keep the event loop from doing anything else.  If the tasks that flush the buffer are in the same event loop, they won't run.  Of course, the asyncio.run above will give it a separate thread to run the loop...

Reply:
Added async version of `wait_for_buffer_flush`